Basic nlohmann-json integration.

axic opened the pull request:
Fixes #6900.

A partial rebase of https://github.com/aarlt/solidity/commit/1329c25756f2b242d9af7860eda6f37a72d38dc0.

aarlt committed:
Basic nlohmann-json integration.

Co-authored-by: Alex Beregszaszi <alex@rtfs.hu>

axic commented on cmake/nlohmann-json.cmake:
```diff
@@ -0,0 +1,14 @@
+include(ExternalProject)
+
+ExternalProject_Add(nlohmann-json
```
Should just git import the file, shouldn't we?

cameel replied:
It's a huge file though (> 25k lines). I don't think that adding external code to the repo is much less of a hack than downloading the package in cmake config.

Maybe we should look into adopting some dependency management system to make using libraries less of a pain? I see that for example [Conan](https://docs.conan.io) and [Hunter](https://hunter.readthedocs.io) are getting some popularity in the last few years. There's also [vcpkg](https://github.com/microsoft/vcpkg) but it seems to be tied to MS stuff and VS in some ways even if it's ultimately meant to be cross-platform. Never used any of them but they look promising.

Tagging @ekpyron becuse he'll likely have an opinion on this :)

ekpyron replied:
**Please** don't pull in crazy package managers in the build system, things are bad enough as they are :-).
We should just use submodules for this stuff and be done with it. I know there have been complications with using submodules in the past, but IIUC that was due to developing both ends in parallel, which is much more of a hassle.
Pulling in fixed versions of external code that are occasionally updated with submodules is zero hassle.

ekpyron replied:
IMHO the only viable alternative to git subtrees or git submodules (of which I'd prefer the latter for this), is just don't do anything at all, document the dependency, have users install it and cmake find it. Everything else is enormously bad and that goes *especially* for "dependency management systems". That's the worst you can do to anyone who ever wants to package your package and *that's* who one should be concerned about primarily - regular users rarely build anyways.

ekpyron replied:
Just to rant a bit more about this :-D: https://github.com/ethereum/solc-bin/pull/101 is exactly what you get from dependency managers  - it's **so** much more convenient for everyone to just use a random dependency manager (which of course doesn't have like an API and interface of its own you first need to know about, if you ever want to override anything), that it's *totally* worth it to then require people to track down opaque dependency chains with 10 indirections without being able to influence any of it. That's *really* superior to the *extreme* hassle of an occasional ``git submodule update`` :-).

But yeah, sorry :-D. I really think the seeming convenience of dependency managers is an absolute fallacy - it exactly prevents you from hitting problems in cases in which you definitely *should* hit problems as in that you should update your stuff to work with proper up-to-date external dependencies. Even submodules are actually bad in that regard, but for some small header-only dependencies they may be excusable...

cameel replied:
I think we should continue this discussion in the context of #8860. I quoted the comments there and will post my response in that issue in a moment.

stackenbotten commented on libdevcore/JSON.cpp:
```diff
@@ -0,0 +1,82 @@
+ * @date 2018
+ */
+
+#include "JSON.h"
```
Coding style error

stackenbotten replied:
Coding style error

stackenbotten commented on libdevcore/JSON.cpp:
```diff
@@ -0,0 +1,82 @@
+				  << "exception id: " << e.id << '\n'
+				  << "byte position of error: " << e.byte << std::endl;
+		*/
+		if (_errs) {
```
Coding style error

stackenbotten replied:
Coding style error

stackenbotten replied:
Coding style error

stackenbotten commented on libdevcore/JSON.cpp:
```diff
@@ -0,0 +1,82 @@
+				  << "exception id: " << e.id << '\n'
+				  << "byte position of error: " << e.byte << std::endl;
+		*/
+		if (_errs) {
```
Coding style error

stackenbotten replied:
Coding style error

stackenbotten replied:
Coding style error

stackenbotten commented on libdevcore/JSON.cpp:
```diff
@@ -0,0 +1,82 @@
+
+using namespace std;
+
+namespace dev {
```
Coding style error

stackenbotten replied:
Coding style error

stackenbotten replied:
Coding style error

stackenbotten commented on libsolutil/JSON.cpp:
```diff
@@ -104,46 +60,46 @@ void removeNullMembersHelper(Json::Value& _json)
+	}
+	catch (Json::parse_error &e) {
+		// NOTE: e.id() gives the code and e.byte() gives the byte offset
+		if (_errs) {
```
Coding style error

stackenbotten commented on libsolutil/JSON.cpp:
```diff
@@ -104,46 +60,46 @@ void removeNullMembersHelper(Json::Value& _json)
+		_json = Json::parse(_input);
+		return true;
+	}
+	catch (Json::parse_error &e) {
```
Coding style error

stackenbotten commented on libsolutil/JSON.cpp:
```diff
@@ -104,46 +60,46 @@ void removeNullMembersHelper(Json::Value& _json)
+	}
+	catch (Json::parse_error &e) {
+		// NOTE: e.id() gives the code and e.byte() gives the byte offset
+		if (_errs) {
```
Coding style error

stackenbotten commented on libsolutil/JSON.cpp:
```diff
@@ -104,46 +60,46 @@ void removeNullMembersHelper(Json::Value& _json)
+		_json = Json::parse(_input);
+		return true;
+	}
+	catch (Json::parse_error &e) {
```
Coding style error

stackenbotten commented on libsolutil/JSON.cpp:
```diff
@@ -104,46 +60,49 @@ void removeNullMembersHelper(Json::Value& _json)
+		_json = Json::parse(_input);
+		return true;
+	}
+	catch (Json::parse_error &e)
```
Coding style error

axic commented on libsolutil/JSON.cpp:
```diff
@@ -104,46 +60,49 @@ void removeNullMembersHelper(Json::Value& _json)
+Json removeNullMembers(Json _json)
 {
-	removeNullMembersHelper(_json);
+	// TODO: Support this.
```
This needs to be fixed.

axic commented on libsolutil/JSON.cpp:
```diff
@@ -104,46 +60,49 @@ void removeNullMembersHelper(Json::Value& _json)
-	if (_format.format == JsonFormat::Pretty)
-		boost::replace_all(result, " \n", "\n");
-	return result;
+	// TODO: support the other features
```
This needs to be fixed.

axic committed:
Remove last remnants of jsoncpp

axic commented on scripts/create_source_tarball.sh:
```diff
@@ -31,7 +31,6 @@ REPO_ROOT="$(dirname "$0")"/..
     fi
     # Add dependencies
     mkdir -p "$SOLDIR/deps/downloads/" 2>/dev/null || true
-    wget -O "$SOLDIR/deps/downloads/jsoncpp-1.9.3.tar.gz" https://github.com/open-source-parsers/jsoncpp/archive/1.9.3.tar.gz
```
Why does this needs jsoncpp, but not range-v3?

`scripts/release_ppa.sh` has both.

cameel replied:
Because it does not need either so it did not fail when we were adding `ranges-v3` and forgot to add it here.

If it's here though, the package gets included in sources and build does not rely on an internet connection. It might have been more important in the past when our cmake config was not downloading jsoncpp automatically - this script was created in #1542 and had the wget line from the very beginning while `jsoncpp.cmake` that downloads jsoncpp was only added in #2926.

Personally, I think that external libs should not be a part of our source package but if we do want that then both libs should be listed here. In that case we should also add a cmake option to disable the download - so that we get an error here if we add more libs in the future and forget to list one of them.

cameel commented on libsolidity/ast/AST.h:
```diff
@@ -32,10 +32,9 @@
 #include <liblangutil/SourceLocation.h>
 #include <libevmasm/Instruction.h>
 #include <libsolutil/FixedHash.h>
+#include <libsolutil/JSON.h>
```
This seems unused in this file.

axic replied:
The first commit started with reusing @aarlt's change. That basically includes `libsolutil/JSON.h` to implicitly include the nlohmann header.

cameel replied:
But I mean that `AST.h` does not seem to have anything JSON related (at least a quick search does not find anything named "json"). So I guess it's either not used at all or used indirectly by other files that include `AST.h`. I think we should have those files include `libsolutil/JSON.h` directly.

axic replied:
Removed, compiles.

cameel commented on libevmasm/Assembly.h:
```diff
@@ -28,12 +28,11 @@
 #include <libsolutil/Common.h>
 #include <libsolutil/Assertions.h>
+#include <libsolutil/JSON.h>
 #include <libsolutil/Keccak256.h>
 
 #include <libsolidity/interface/OptimiserSettings.h>
 
-#include <json/json.h>
```
Including the library via our own header to abstract it out a bit seems like a good idea but on the other hand I wonder how it would interact with `include-what-you-use` if we were to start using it (#11813).

axic replied:
Just commented a relevant piece here https://github.com/ethereum/solidity/pull/11967#discussion_r715829063

cameel commented on libsolidity/ast/ASTJsonConverter.cpp:
```diff
@@ -130,33 +130,33 @@ string ASTJsonConverter::namePathToString(std::vector<ASTString> const& _namePat
 	}
 	else
-		return Json::nullValue;
+		return Json{};
```
So there's no explicit `null` value in this lib?

EDIT: I see in your tests that `nullptr` is treated as `null`. I guess we don't want to use it in that role though?

axic replied:
Yes, the commits are kind of organically developed, since I added those tests we can decide what style we want to use.

cameel replied:
I think `Json{}` or `{}` is fine. `nullptr` would look a bit like `null` but on the other hand I can't not see it as a pointer :) It's just interesting that the lib accepts it as `null`.

cameel commented on libsolidity/ast/ASTJsonConverter.cpp:
```diff
@@ -171,16 +171,17 @@ void ASTJsonConverter::appendExpressionAttributes(
-		tuple["suffix"] = Json::Value(_info.second.suffix);
-	tuple["valueSize"] = Json::Value(Json::LargestUInt(_info.second.valueSize));
+		tuple["suffix"] = Json(_info.second.suffix);
+	// TODO: assert?
+	tuple["valueSize"] = Json(static_cast<Json::number_integer_t>(_info.second.valueSize));
```
Yeah, I'd definitely assert. This can truncate the value if it's too big, which I don't think we'd be ok with happening.

Also, is the range of `Json::LargestUInt` and `Json::number_integer_t` the same?

axic replied:
I think we actually want to review this whole mess with jsoncpp -- decide on a single number type and have assertions. Probably need a helper for that.

Then it is simple to keep using that with nlohmann-json.

cameel commented on libsolidity/ast/ASTJsonImporter.cpp:
```diff
@@ -72,11 +72,11 @@ map<string, ASTPointer<SourceUnit>> ASTJsonImporter::jsonToSourceUnit(map<string
+ASTPointer<T> ASTJsonImporter::createASTNode(Json const& _node, Args&&... _args)
 {
-	astAssert(member(_node, "id").isInt64(), "'id'-field must be 64bit integer.");
+	astAssert(member(_node, "id").is_number_integer(), "'id'-field must be 64bit integer.");
```
[`integer_t`](https://nlohmann.github.io/json/api/basic_json/number_integer_t/) seems to be a template parameter. It defaults to `int64_t` but I'd use `static_assert` to document that it must be that eaxact type here.

axic replied:
Added a `static_assert`.

cameel commented on libsolidity/ast/ASTJsonImporter.cpp:
```diff
@@ -915,129 +915,129 @@ ASTPointer<IndexRangeAccess> ASTJsonImporter::createIndexRangeAccess(Json::Value
 {
-	if (!_node.isMember(_name))
-		return Json::nullValue;
+	if (_node.find(_name) == _node.end())
+		return Json{};
```
I think that [`contains()`](https://nlohmann.github.io/json/api/basic_json/contains/) would be a better replacement for `isMember()`.

axic replied:
Yes, I found it after doing some of these and fixed most places, but probably missed some.

cameel replied:
I've seen quite a few, but for now pointed out only some of them. I can point out the rest in the second pass.

cameel commented on libsolidity/interface/CompilerStack.cpp:
```diff
@@ -1454,15 +1452,15 @@ string CompilerStack::createMetadata(Contract const& _contract, bool _forIR) con
 	}
 
-	static_assert(sizeof(m_optimiserSettings.expectedExecutionsPerDeployment) <= sizeof(Json::LargestUInt), "Invalid word size.");
-	solAssert(static_cast<Json::LargestUInt>(m_optimiserSettings.expectedExecutionsPerDeployment) < std::numeric_limits<Json::LargestUInt>::max(), "");
```
This assert seems wrong. There should be no cast there. If we cast to `LargestUInt`, we'll always get a value lower than max for this type :)

cameel commented on libsolidity/interface/CompilerStack.cpp:
```diff
@@ -1454,15 +1452,15 @@ string CompilerStack::createMetadata(Contract const& _contract, bool _forIR) con
-	static_assert(sizeof(m_optimiserSettings.expectedExecutionsPerDeployment) <= sizeof(Json::LargestUInt), "Invalid word size.");
-	solAssert(static_cast<Json::LargestUInt>(m_optimiserSettings.expectedExecutionsPerDeployment) < std::numeric_limits<Json::LargestUInt>::max(), "");
-	meta["settings"]["optimizer"]["runs"] = Json::Value(Json::LargestUInt(m_optimiserSettings.expectedExecutionsPerDeployment));
+	static_assert(sizeof(m_optimiserSettings.expectedExecutionsPerDeployment) <= sizeof(Json::number_integer_t), "Invalid word size.");
```
Wouldn't it be better to compare `std::numeric_limits<T>::max()` of both types here? A `sizeof()` check will still pass if `expectedExecutionsPerDeployment` is e.g. `uint64_t` while the limit is `int64_t`.

axic replied:
There seems to be a general mess regarding (large) numbers, both currently and with nlohmann.

cameel commented on test/libsolutil/JSON.cpp:
```diff
@@ -90,12 +86,12 @@ BOOST_AUTO_TEST_CASE(json_pretty_print)
 	"  {\n"
-	"    \"3.1\": \"3.1\",\n"
-	"    \"3.2\": 2\n"
+	"    \"3.1\" : \"3.1\",\n"
+	"    \"3.2\" : 2\n"
```
Does the test fail without the extra spaces before `:`? Seems like a weird limitation.

cameel commented on test/libsolutil/JSON.cpp:
```diff
@@ -135,31 +131,28 @@ BOOST_AUTO_TEST_CASE(parse_json_strict)
+	BOOST_CHECK(jsonParseStrict("null", json, &errors));
+	BOOST_CHECK(json.is_null());
 
 	// Single quotes are also disallowed by ECMA-404
 	BOOST_CHECK(!jsonParseStrict("'hello'", json, &errors));
```
So this part is non-compliant? Do we need it to be compliant?

axic replied:
This seems to be properly handled as failure, could add the `error` message check here though.

cameel commented on test/libsolutil/JSON.cpp:
```diff
@@ -172,15 +165,12 @@ BOOST_AUTO_TEST_CASE(parse_json_strict)
+	BOOST_CHECK(errors == "[json.exception.parse_error.101] parse error at line 1, column 4: syntax error while parsing value - invalid string: ill-formed UTF-8 byte; last read: '\"?'");
 
 	// This would be valid more lenient parsers.
 	BOOST_CHECK(jsonParseStrict("[ \"\xF0\x9F\x98\x8A\" ]", json, &errors));
-	BOOST_CHECK(json.isArray());
+	BOOST_CHECK(json.is_array());
 	BOOST_CHECK(json[0] == "😊");
```
This too.

cameel commented on test/libsolidity/StandardCompiler.cpp:
```diff
@@ -55,97 +63,98 @@ langutil::Error::Severity str2Severity(string const& _cat)
 		{
-			BOOST_TEST_REQUIRE(linkReferenceResult[fileName][libraryName].isArray());
+			BOOST_TEST_REQUIRE(linkReferenceResult[fileName][libraryName].is_array());
 			BOOST_TEST_REQUIRE(!linkReferenceResult[fileName][libraryName].empty());
-			for (int i = 0; i < static_cast<int>(linkReferenceResult.size()); ++i)
-			{
-				BOOST_TEST_REQUIRE(linkReferenceResult[fileName][libraryName][i].isObject());
-				BOOST_TEST_REQUIRE(linkReferenceResult[fileName][libraryName][i].size() == 2);
-				BOOST_TEST_REQUIRE(linkReferenceResult[fileName][libraryName][i]["length"].isUInt());
-				BOOST_TEST_REQUIRE(linkReferenceResult[fileName][libraryName][i]["start"].isUInt());
-			}
+// TODO: rewrite this properly
+//			for (int i = 0; i < static_cast<int>(linkReferenceResult.size()); ++i)
+//			{
+//				BOOST_TEST_REQUIRE(linkReferenceResult[fileName][libraryName][i].is_object());
+//				BOOST_TEST_REQUIRE(linkReferenceResult[fileName][libraryName][i].size() == 2);
+//				BOOST_TEST_REQUIRE(linkReferenceResult[fileName][libraryName][i]["length"].is_number_unsigned());
+//				BOOST_TEST_REQUIRE(linkReferenceResult[fileName][libraryName][i]["start"].is_number_unsigned());
+//			}
```
What's the problem with this one?

axic replied:
I don't remember 😅

cameel commented on test/libsolidity/StandardCompiler.cpp:
```diff
@@ -55,97 +63,98 @@ langutil::Error::Severity str2Severity(string const& _cat)
+	BOOST_TEST_REQUIRE(linkReferenceResult.is_object());
 
-	for (string const& fileName: linkReferenceResult.getMemberNames())
+	for (auto const& [fileName, _]: linkReferenceResult.items())
```
Would this work?
```suggestion
	for (string const& fileName: linkReferenceResult.items() | ranges::views::keys)
```

axic replied:
Unfortunately `ranges::views` is not compatible with `items()`.

cameel commented on test/libsolidity/StandardCompiler.cpp:
```diff
@@ -41,6 +41,14 @@ namespace solidity::frontend::test
 namespace
 {
 
+set<string> getMemberNames(Json const& _json)
+{
+	set<string> keys;
+	for (auto const& [key, _]: _json.items())
+		keys.insert(key);
+	return keys;
+}
```
Seems unused. Is it meant to replace the `.getMemberNames()` method?

axic replied:
It supposed to be used in one function, where it was messy doing it otherwise.

cameel replied:
ok

cameel commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -425,7 +425,7 @@ TestCase::TestResult SemanticTest::runTest(
 				soltestAssert(
 					m_allowNonExistingFunctions ||
-					m_compiler.methodIdentifiers(m_compiler.lastContractName(m_sources.mainSourceFile)).isMember(test.call().signature),
+					m_compiler.methodIdentifiers(m_compiler.lastContractName(m_sources.mainSourceFile)).count(test.call().signature),
```
```suggestion
					m_compiler.methodIdentifiers(m_compiler.lastContractName(m_sources.mainSourceFile)).contains(test.call().signature),
```

Also in other files.

axic replied:
Done.

cameel commented on test/libsolidity/GasTest.cpp:
```diff
@@ -123,14 +120,14 @@ TestCase::TestResult GasTest::run(ostream& _stream, string const& _linePrefix, b
 		return estimates.size() == expectations.second.size() &&
 			boost::all(expectations.second, [&](auto const& entry) {
-				return entry.second == estimates[entry.first].asString();
+				return entry.second == estimates[entry.first].template get<string>();
```
Wouldn't it be better to just declare `estimates` as `Json const&` rather than `auto const&`?

cameel commented on solc/CommandLineInterface.cpp:
```diff
@@ -463,23 +463,23 @@ bool CommandLineInterface::readInputFiles()
+		astAssert(ast.contains("sources"), "Invalid Format for import-JSON: Must have 'sources'-object");
 
-		for (auto& src: ast["sources"].getMemberNames())
+		for (auto const& [src, _]: ast["sources"].items())
```
```suggestion
		for (auto const& src: ast["sources"].items() | ranges::views::keys)
```

cameel commented on libsolutil/JSON.h:
```diff
@@ -23,15 +23,17 @@
 
 #include <string>
 
+using Json = nlohmann::json;
```
Do we really want this in a header?

axic replied:
I don't think so, grandfathered in from @aarlt.

Marenz replied:
Implying that all headers should use the full notation?

cameel replied:
Yeah, pretty much. I don't think we want to pollute the global namespace this way, do we?

Marenz replied:
I suspect it will be very verbose to always write `nlohmann::json` instead of `Json`. It's also not exactly rolling of the tongue. 
Also, doesn't this conflict with the desire to abstract the library specific parts as mentioned here https://github.com/ethereum/solidity/pull/11967#discussion_r715772723

Marenz replied:
And I guess it also means we'd need to use `using Json = nlohmann::json;` in all cpp files that use it then?

cameel commented on libsolutil/CMakeLists.txt:
```diff
@@ -42,7 +42,7 @@ set(sources
 
 add_library(solutil ${sources})
-target_link_libraries(solutil PUBLIC jsoncpp Boost::boost Boost::filesystem Boost::system range-v3)
+target_link_libraries(solutil PUBLIC Boost::boost Boost::filesystem Boost::system range-v3)
```
So the new lib is header only. And it's a massive header. How does this affect compilation times?

axic replied:
Good question.

falbrechtskirchinger replied:
FYI, there's a forward declaration header (`json_fwd.hpp`) that can help with compile times. Since v3.11.x, there's also the added benefit of getting linker errors instead of incorrect behavior or crashes at runtime when inappropriately mixing different library versions or configurations.

cameel commented on libsolidity/interface/CompilerStack.cpp:
```diff
@@ -1645,34 +1643,35 @@ bytes CompilerStack::createCBORMetadata(Contract const& _contract, bool _forIR)
 	else
-		return Json::Value(util::toString(_gas.value));
+		return Json(util::toString(_gas.value));
+//		return Json(_gas.value.str());
```
Commented-out code.

axic replied:
I think this meant to be kept as a TODO to double check stuff. Will replace it with an actual TODO.

cameel commented on libsolidity/interface/StandardCompiler.cpp:
```diff
@@ -49,65 +49,65 @@ using namespace solidity::langutil;
 	string const& _formattedMessage = "",
-	Json::Value const& _sourceLocation = Json::Value(),
-	Json::Value const& _secondarySourceLocation = Json::Value()
+	Json const& _sourceLocation ),
+	Json const& _secondarySourceLocation )
```
```suggestion
	Json const& _sourceLocation,
	Json const& _secondarySourceLocation
```

cameel commented on libsolidity/interface/StandardCompiler.cpp:
```diff
@@ -146,13 +146,13 @@ Json::Value formatErrorWithException(
 	{
 		string key = (sourceName == "*") ? "" : sourceName;
-		for (auto const& contractName: _outputSelection[sourceName].getMemberNames())
+		for (auto const& [contractName, _]: contractList.items())
```
```suggestion
		for (string const& contractName: contractList.items() | ranges::views::keys)
```

cameel commented on libsolidity/interface/StandardCompiler.cpp:
```diff
@@ -208,13 +208,13 @@ bool isArtifactRequested(Json::Value const& _outputSelection, string const& _art
 
 	for (auto const& file: { _file, string("*") })
-		if (_outputSelection.isMember(file) && _outputSelection[file].isObject())
+		if (_outputSelection.find(file) != _outputSelection.end() && _outputSelection[file].is_object())
```
```suggestion
		if (_outputSelection.contains(file) && _outputSelection[file].is_object())
```

axic replied:
Done.

cameel commented on libsolidity/interface/StandardCompiler.cpp:
```diff
@@ -557,32 +553,32 @@ std::optional<Json::Value> checkOutputSelection(Json::Value const& _outputSelect
+		if (!_jsonInput["runs"].is_number_unsigned())
 			return formatFatalError("JSONError", "The \"runs\" setting must be an unsigned number.");
-		settings.expectedExecutionsPerDeployment = _jsonInput["runs"].asUInt();
+		settings.expectedExecutionsPerDeployment = _jsonInput["runs"].get<Json::number_unsigned_t>();
```
We should validate the range of the value or have a static assert that types are the same.

cameel commented on libsolidity/interface/StandardCompiler.cpp:
```diff
@@ -740,67 +736,69 @@ std::variant<StandardCompiler::InputsAndSettings, Json::Value> StandardCompiler:
-	Json::Value const& settings = _input.get("settings", Json::Value());
+//	if (!_input.contains("settings"))
+//		_input["settings"] = Json::object();
+	Json const& settings = _input["settings"];
```
So there's no `get()` with a default?

axic replied:
Yes, that's one of the questions how to do that nicely.

cameel replied:
We could just create a helper function for that.

axic replied:
There's actually `.value(key, default)` instead. Used that now in most places.

cameel commented on libsolidity/interface/StandardCompiler.cpp:
```diff
@@ -1513,24 +1514,24 @@ string StandardCompiler::compile(string const& _input) noexcept
 		else
-			fun["id"] = Json::nullValue;
+			fun["id"] = Json{};
 		if (info.bytecodeOffset)
-			fun["entryPoint"] = Json::UInt64(*info.bytecodeOffset);
+			fun["entryPoint"] = Json::number_unsigned_t(*info.bytecodeOffset);
 		else
-			fun["entryPoint"] = Json::nullValue;
-		fun["parameterSlots"] = Json::UInt64(info.params);
-		fun["returnSlots"] = Json::UInt64(info.returns);
+			fun["entryPoint"] = Json{};
+		fun["parameterSlots"] = Json::number_unsigned_t(info.params);
+		fun["returnSlots"] = Json::number_unsigned_t(info.returns);
```
These also needs some (static) asserts about sizes.

axic replied:
Are slots actually large values?

cameel commented on libsolutil/JSON.cpp:
```diff
@@ -33,8 +33,8 @@
 
 static_assert(
-	(JSONCPP_VERSION_MAJOR == 1) && (JSONCPP_VERSION_MINOR == 9) && (JSONCPP_VERSION_PATCH == 3),
-	"Unexpected jsoncpp version: " JSONCPP_VERSION_STRING ". Expecting 1.9.3."
+	(NLOHMANN_JSON_VERSION_MAJOR == 3) && (NLOHMANN_JSON_VERSION_MINOR == 10) && (NLOHMANN_JSON_VERSION_PATCH == 2),
+	"Unexpected nlohmann-json version. Expecting 3.10.2."
 );
```
Do we actually need to enforce one specific version? Why not allow a range?

axic replied:
Not sure, just kept the same logic.

axic replied:
Well we download a specific file too, so this is just a safety check. I think it is fine.

cameel commented on libsolutil/JSON.cpp:
```diff
@@ -104,46 +60,53 @@ void removeNullMembersHelper(Json::Value& _json)
+	try
+	{
+		_json = Json::parse(_input);
+		return true;
```
Shouldn't we also set `_errs` to an empty string if there are no errors?

axic replied:
I thought about that. I kind of think there's still a second cleanup PR we could based on the learnings here, just to fix/improve things in general. This change is not specific to nlohmann-json.

cameel replied:
Yeah. Ok, leaving this comment open so that we can still find it when doing that second PR.

cameel commented on libsolutil/JSON.cpp:
```diff
@@ -104,46 +60,53 @@ void removeNullMembersHelper(Json::Value& _json)
 }
 
-bool jsonParseStrict(string const& _input, Json::Value& _json, string* _errs /* = nullptr */)
+bool jsonParseStrict(string const& _input, Json& _json, string* _errs /* = nullptr */)
```
Not necessarily in this PR but I think it would be better if this was returning either `optional<Json>` or maybe `variant<Json, string>`.

axic commented:
@cameel any idea why the external project doesn't on emscripten and arch?

cameel commented:
> @cameel any idea why the external project doesn't work on emscripten and arch?

It actually seems to fail everywhere (except for Windows for some reason). In case of emscripten it seems to be because it needs `emcmake` in cmake command. Both ranges and jsoncpp set `CMAKE_COMMAND` for that reason.

I don't know why other platforms fail but it might have something to do with you setting `CMAKE_ARGS` simply to `true`. These other libs have an extensive set of parameters for it.

By the way, maybe it would be better to use `FetchContent` instead as in #11273?

axic commented:
Unrelated, but we also have this warning:
```
[  1%] No update step for 'range-v3-project'
[  1%] Built target evmc_loader
[  2%] No patch step for 'range-v3-project'
[  2%] Performing configure step for 'range-v3-project'
-- The CXX compiler identification is AppleClang 12.0.5.12050022
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Warning at cmake/ranges_env.cmake:77 (message):
  [range-v3 warning]: unknown build type, defaults to release!
Call Stack (most recent call first):
  CMakeLists.txt:50 (include)


-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE) 
-- Found Git: /usr/bin/git (found version "2.30.1 (Apple Git-130)") 
```

Notice `[range-v3 warning]: unknown build type, defaults to release!`.

axic committed:
Use FetchContent for nlohmann

axic commented:
> By the way, maybe it would be better to use FetchContent instead as in #11273?

@cameel I actually did `FetchContent` in 5510abd, but it had the same problem.

I'd refrain from rebasing this before the problem is solved though.

axic committed:
Import nhlomann/json

It is the single_include/json.hpp file from https://github.com/nlohmann/json/releases/tag/v3.10.2

axic commented on libsolidity/interface/ABI.cpp:
```diff
@@ -96,79 +98,79 @@ Json::Value ABI::generate(ContractDefinition const& _contractDef)
 
-	Json::Value abiJson{Json::arrayValue};
+	Json abiJson{Json::array()};
 	for (auto& f: abi)
```
Is the point of this to enforce the ordering?

axic commented on libsolidity/interface/StandardCompiler.cpp:
```diff
@@ -903,15 +898,18 @@ std::variant<StandardCompiler::InputsAndSettings, Json::Value> StandardCompiler:
 
-	ret.metadataLiteralSources = metadataSettings.get("useLiteralContent", Json::Value(false)).asBool();
-	if (metadataSettings.isMember("bytecodeHash"))
+	ret.metadataLiteralSources =
```
Could use `metadataSettings.value("useLiteralContent", false)`.

axic commented on libsolutil/JSON.cpp:
```diff
@@ -43,53 +43,9 @@ namespace solidity::util
+#if 0
 /// Takes a JSON value (@ _json) and removes all its members with value 'null' recursively.
-void removeNullMembersHelper(Json::Value& _json)
+void removeNullMembersHelper(Json& _json)
```
TODO: need to check if this needed, and if yes, implement it.

axic commented on test/libsolidity/StandardCompiler.cpp:
```diff
@@ -41,6 +41,14 @@ namespace solidity::frontend::test
 namespace
 {
 
+set<string> getMemberNames(Json const& _json)
```
Ideally instead of this we would use range-v3 to get the keys from the `items()` iterator, but haven't found a nice way yet.

leonardoalt commented:
I guess this needs to be taken over?

axic commented:
> I guess this needs to be taken over?

I think this ran into some bug in one of the JSON output generators.

But also let's wait for https://github.com/nlohmann/json/issues/3130 to be fixed. (Well was just merged now.)

falbrechtskirchinger commented:
> > I guess this needs to be taken over?
> 
> I think this ran into some bug in one of the JSON output generators.
> 
> But also let's wait for [nlohmann/json#3130](https://github.com/nlohmann/json/issues/3130) to be fixed. (Well was just merged now.)

Let me know if there're any more issues. I've not explicitly tested with range-v3 but `std::ranges`. I expect 3.11.0 to be released soon (although it's not up to me).

Marenz commented:
I pushed a rebase and will now look & go through the comments here

axic committed:
Import nhlomann/json

From https://github.com/nlohmann/json/releases/download/v3.11.2/json.hpp

axic commented:
Rebased and upgraded to nlohmann-json 3.11.2 which has proper ranges-v3 support.

axic committed:
Fix typo in nlohmann-json

axic committed:
Some review fixes

axic committed:
Different assert

axic committed:
remove jsoncp from osx ci

axic committed:
enable diag

aarlt commented on libsolutil/JSON.cpp:
```diff
@@ -104,46 +60,54 @@ void removeNullMembersHelper(Json::Value& _json)
+	return _input.dump(
+		/* indent */ (_format.format == JsonFormat::Pretty) ? static_cast<int>(_format.indent) : -1,
+		/* indent_char */ ' ',
+		/* ensure_ascii */ true
```
should be `false` here - we have some tests using unicode

axic replied:
This means unicode stuff is escaped, which is the default behaviour of JSON expected by many implementations.

aarlt replied:
I just noticed that some tests are not working because of that. But yeah, we could also rewrite the tests - but maybe we should not.

axic committed:
Revert "[libsolutil] Add new JSON helper functions."

This reverts commit c23093e0f0a28bcc6a05fe06953d70be3f96cee3.

axic committed:
more conversions

axic committed:
update tests: abijson

axic committed:
Updates

ekpyron commented:
Closing this as "postponed" for now. We'll keep the branch around as a basis for when we may pick this up again.
